File: client/java-client/src/main/java/com/kotak/neo/client/api/MarginApi.java
package com.kotak.neo.client.api;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.kotak.neo.client.Configuration;
import com.kotak.neo.client.Settings;
import com.kotak.neo.client.rest.RestClient;
import com.kotak.neo.client.validation.Validators;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;

public class MarginApi {
    private final RestClient rest;
    private final Configuration config;

    public MarginApi(RestClient rest, Configuration config) {
        this.rest = rest;
        this.config = config;
    }

    public JsonElement marginRequired(MarginRequiredRequest r) {
        Validators.validateMargin(r.exchangeSegment, r.price, r.orderType, r.product,
                r.quantity, r.instrumentToken, r.transactionType);

        Map<String, Object> body = new LinkedHashMap<>();
        body.put("exSeg", Settings.EXCHANGE_SEGMENT.get(r.exchangeSegment));
        body.put("prc", r.price);
        body.put("prcTp", Settings.ORDER_TYPE.get(r.orderType));
        body.put("prod", Settings.PRODUCT.get(r.product));
        body.put("qty", r.quantity);
        body.put("tok", r.instrumentToken);
        body.put("trnsTp", r.transactionType);
        body.put("trgPrc", r.triggerPrice);
        body.put("brkName", r.brokerName == null ? "KOTAK" : r.brokerName);
        body.put("brnchId", r.branchId == null ? "ONLINE" : r.branchId);
        body.put("slAbsOrTks", r.stopLossType);
        body.put("slVal", r.stopLossValue);
        body.put("sqrOffAbsOrTks", r.squareOffType);
        body.put("sqrOffVal", r.squareOffValue);
        body.put("trailSL", r.trailingStopLoss);
        body.put("tSLTks", r.trailingSLValue);

        Map<String, String> headers = new HashMap<>();
        headers.put("Sid", config.editSid == null ? "" : config.editSid);
        headers.put("Auth", config.editToken == null ? "" : config.editToken);
        headers.put("Content-Type", "application/x-www-form-urlencoded");

        Map<String, String> query = new HashMap<>();
        query.put("sId", config.serverId == null ? "" : config.serverId);

        JsonElement resp = rest.request("POST", config.getUrl("margin"), query, headers, body).data;
        JsonObject wrap = new JsonObject();
        wrap.add("data", resp);
        return wrap;
    }

    public static class MarginRequiredRequest {
        public String exchangeSegment, price, orderType, product, quantity, instrumentToken, transactionType;
        public String triggerPrice, brokerName, branchId, stopLossType, stopLossValue,
                squareOffType, squareOffValue, trailingStopLoss, trailingSLValue;
    }
}

File: client/typescript-client/src/api/marginApi.ts
import { Configuration } from "../configuration";
import { RestClient, Response } from "../rest";
import { EXCHANGE_SEGMENT, ORDER_TYPE, PRODUCT } from "../settings";
import { MarginRequiredRequest } from "../types";
import { validateMargin } from "../validation";

export class MarginApi {
  constructor(private rest: RestClient, private config: Configuration) {}

  async marginRequired(r: MarginRequiredRequest): Promise<Response> {
    validateMargin(r);
    const res = await this.rest.request({
      method: "POST",
      url: this.config.getUrl("margin"),
      queryParams: { sId: this.config.serverId || "" },
      headers: {
        Sid: this.config.editSid || "",
        Auth: this.config.editToken || "",
        "Content-Type": "application/x-www-form-urlencoded",
      },
      body: {
        exSeg: EXCHANGE_SEGMENT[r.exchange_segment],
        prc: r.price,
        prcTp: ORDER_TYPE[r.order_type],
        prod: PRODUCT[r.product],
        qty: r.quantity,
        tok: r.instrument_token,
        trnsTp: r.transaction_type,
        trgPrc: r.trigger_price,
        brkName: r.broker_name ?? "KOTAK",
        brnchId: r.branch_id ?? "ONLINE",
        slAbsOrTks: r.stop_loss_type,
        slVal: r.stop_loss_value,
        sqrOffAbsOrTks: r.square_off_type,
        sqrOffVal: r.square_off_value,
        trailSL: r.trailing_stop_loss,
        tSLTks: r.trailing_sl_value,
      },
    });
    return { data: res.data };
  }
}
